qiskit/backends/_sympy_qasmsimulator.py

```diff
@@ -0,0 +1,438 @@
+
+        pdist = [SympyQasmSimulator._conjugate_square(matrix_form[i, 0]) for i in range(shapeN)]
+        norm_pdist = [float(i)/sum(pdist) for i in pdist]
+
+        for i in range(actual_shots):  # pylint: disable=unused-variable
+            _classical_state_observed = np.random.choice(np.arange(0, shapeN), p=norm_pdist)
+            outcomes.append(bin(_classical_state_observed)[2:].zfill(
+                self._number_of_cbits))
+
+        # Return the results
+        counts = dict(Counter(outcomes))
+        data = {'counts': self._format_result(counts, cl_reg_index, cl_reg_nbits)}
+
+        data['quantum_state'] = np.asarray(list_form)  # consistent with other backends
+        data['classical_state'] = self._classical_state
+        return {'data': data, 'status': 'DONE'}
+
+    def _format_result(self, counts, cl_reg_index, cl_reg_nbits):
+        """Format the result bit string.
+        This formats the result bit strings such that spaces are inserted
+        at register divisions.
+
+        Args:
+            counts (dict): dictionary of counts e.g. {'1111': 1000, '0000':5}
+            cl_reg_index (list): starting bit index of classical register
+            cl_reg_nbits (list): total amount of bits in classical register
+        Returns:
+            dict: spaces inserted into dictionary keys at register boundaries.
+        """
+        fcounts = {}
+        for key, value in counts.items():
+            new_key = [key[-cl_reg_nbits[0]:]]
+            for index, nbits in zip(cl_reg_index[1:],
+                                    cl_reg_nbits[1:]):
+                new_key.insert(0, key[-(index+nbits):-index])
+            fcounts[' '.join(new_key)] = value
+        return fcounts
+
+    @staticmethod
+    def regulate(theta):
+        """find the symbolic form that is closest to the numeric value"""
+        error_margin = 0.01
+
+        if abs(N(theta - pi)) < error_margin:
+            return pi, True
+        if abs(N(theta - pi/2)) < error_margin:
+            return pi/2, True
+        if abs(N(theta - pi/4)) < error_margin:
+            return pi/4, True
+        if abs(N(theta - 2*pi)) < error_margin:
+            return 2*pi, True
+
+        return theta, theta == 0  # if theta ==0, we also think it is regular
+
+    class SDGGate(OneQubitGate):  # pylint: disable=too-many-ancestors, abstract-method
+        """implements the SDG gate"""
+        gate_name = 'SDG'
+
+        def get_target_matrix(self, format='sympy'):  # pylint: disable=redefined-builtin
+            return Matrix([[1, 0], [0, -I]])
+
+    class TDGGate(OneQubitGate):  # pylint: disable=too-many-ancestors, abstract-method
+        """implements the TDG gate"""
+        gate_name = 'TDG'
+
+        def get_target_matrix(self, format='sympy'):  # pylint: disable=redefined-builtin
+            return Matrix([[1, 0], [0, exp(-I*pi/4)]])
+
+    @staticmethod
+    def compute_ugate_matrix(parafloatlist):
```

Comment:
regulate and compute_ugate_matrix returns run-dependant types (sometimes floats, sometimes symbols), which is not ideal.
In addition a single "inregulateable" value creates a matrix with without symbols.
I think this way is better:
```
def regulate(theta):
    """Find the most close symbolic form"""
    error_margin = 0.01
    targets = [pi, pi/2, pi * 2, pi / 4]

    for t  in targets:
        if abs(N(theta - t)) < error_margin:
            return t

    return sympy.sympify(theta)
```
```
def compute_ugate_matrix(parafloatlist):
    """compute the ugate matrix
        Args:
            parafloatlist(list): parafloatlist
        Returns:
            Matrix: the matrix that represents the U gate
    """
    theta = SympyUnitarySimulator.regulate(parafloatlist[0])
    phi = SympyUnitarySimulator.regulate(parafloatlist[1])
    lamb = SympyUnitarySimulator.regulate(parafloatlist[2])

    left_up = cos(theta/2)
    right_up = (-E**(I*lamb)) * sin(theta/2)
    left_down = (E**(I*phi)) * sin(theta/2)
    right_down = (E**(I*(phi+lamb))) * cos(theta/2)
    return Matrix([[left_up, right_up], [left_down, right_down]])
```

Reply:
I do not see much changes except for the difference in the coding style.

The minor difference I saw is that you used sympy.sympify(), but it can still return the float values.  try sympy.sympify(0.5), it would return 0.500000000000000.

To address your comment, I will wrap theta with sympy.sympify(). 

The coding style is subjective, I will leave it unchanged unless pylint raises alarms.

Reply:
sympify converts to sympy objects:
```
In [3]: type(sympy.sympify(0.5))
Out[3]: sympy.core.numbers.Float
```
The current code returns different types (a regulated theta is sympy object and a non-regulated theta is whatever was originally). It's good practice to always return the same type.